File: exercism/javascript/complete/perfect-numbers/perfect-numbers.js

// find factors including 1 but not itself, sum factors, return -1,0,1, add errors
const PerfectNumbers = () => {
  const classify = (input) => {
    if (input < 1) {
      throw(new Error('Classification is only possible for natural numbers.'));
    } else if (input == 1) {
      return "deficient";
    }
    let factors = [];
    for (let i = 1; i <= input/2; i++) {
      if (input % i == 0) {
        factors.push(i);
      }
    }
    let total = factors.reduce((sum, value) => sum + value);
    if (total == input) {
      return "perfect";
    } else if (total < input) {
      return "deficient";
    } else {
      return "abundant";
    }
  }
  return { classify }
}

export default PerfectNumbers;

File: exercism/ruby/complete/perfect-numbers/perfect_numbers.rb
class PerfectNumber
  def self.classify(number)
    raise RuntimeError if number < 0
    factors = []
    (1..number/2).each do |factor|
      factors << factor if number % factor == 0
    end

    sum = factors.sum
    if sum > number
      'abundant'
    elsif sum < number
      'deficient'
    elsif sum == number
      'perfect'
    end
  end
end
